planning/trajectory_optimization/gcs_trajectory_optimization.h

```diff
@@ -0,0 +1,284 @@
+#pragma once
+
+#include <map>
+#include <memory>
+#include <string>
+#include <tuple>
+#include <utility>
+#include <vector>
+
+#include "drake/common/trajectories/bezier_curve.h"
+#include "drake/common/trajectories/composite_trajectory.h"
+#include "drake/geometry/optimization/convex_set.h"
+#include "drake/geometry/optimization/graph_of_convex_sets.h"
+
+namespace drake {
+namespace planning {
+namespace trajectory_optimization {
+
+/**
+GCSTrajectoryOptimization implements a simplified motion planning optimization
+problem introduced in the paper "Motion Planning around Obstacles with Convex
+Optimization."
+
+"Motion Planning around Obstacles with Convex Optimization" by Tobia Marcucci,
+Mark Petersen, David von Wrangel, Russ Tedrake. https://arxiv.org/abs/2205.04422
+
+Instead of using the full time-scaling curve, this problem uses a single
+time-scaling variable for each region. This formulation can not support convex
+constraints on continuity greater than order 0 on the trajectory, but can
+support continuity on the path for arbitrary degree. The resulting trajectories
+can be post-processed with e.g. Toppra in order to smooth out the timing
+rescaling.
+
+The trajectory is defined as q(t) = r(h(t)).
+r(s) is the path, parametrized as a Bézier curve.
+h(s) = duration * s is the time-scaling function.
+
+This class supports the notion of a Subgraph of regions. This has proven useful
+to facilitate multi-modal motion planning such as: Subgraph A: find a
+collision-free trajectory for the robot to a grasping posture, Subgraph B: find
+a collision-free trajectory for the robot with the object in its hand to a
+placing posture, etc.
+*/
+class GCSTrajectoryOptimization {
+ public:
+  DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(GCSTrajectoryOptimization);
+
+  ~GCSTrajectoryOptimization() = default;
+
+  /**
+  Constructs the motion planning problem.
+  @param num_positions is the dimension of the configuration space.
+  */
+  explicit GCSTrajectoryOptimization(int num_positions);
+
+  class Subgraph final {
```

Comment:
I'm surprised to see that the `Subgraph` class is in the trajectory optimization code. Probably that's because I don't fully understand the moving parts here, but I would have imagined that subgraphs should be defined in terms of the GraphOfConvexSets?